go/worker/storage/committee/checkpoint_sync.go

```diff
@@ -0,0 +1,268 @@
+		case errors.Is(result.err, checkpoint.ErrChunkCorrupted):
+			chunkReturnCh <- chunk
+			return
+		case errors.Is(result.err, checkpoint.ErrChunkProofVerificationFailed):
+			errorCh <- checkpointStatusNext
+			return
+		case result.err != nil:
+			n.logger.Error("chunk restoration failed", "node", clientConn.Node.ID, "chunk", chunk.Index, "err", result.err)
+			return
+		}
+	}
+}
+
+func (n *Node) handleCheckpoint(check *checkpoint.Metadata, committeeClient committee.Client) (int, error) {
+	workCtx, workCancel := context.WithCancel(n.ctx)
+	defer workCancel()
+
+	chunkDispatchCh := make(chan *checkpoint.ChunkMetadata)
+	defer close(chunkDispatchCh)
+
+	// Start a worker for each node in the committee.
+	conns := committeeClient.GetConnectionsWithMeta()
+	if len(conns) == 0 {
+		return checkpointStatusBail, errors.New("no connections to storage committee")
+	}
+
+	chunkReturnCh := make(chan *checkpoint.ChunkMetadata, len(conns))
+	errorCh := make(chan int, len(conns))
+
+	var workerGroup sync.WaitGroup
+	workersDone := make(chan interface{})
+
+	for _, clientConn := range conns {
+		workerGroup.Add(1)
+		go func(clientConn *committee.ClientConnWithMeta) {
+			defer workerGroup.Done()
+			n.nodeWorker(workCtx, clientConn, chunkDispatchCh, chunkReturnCh, errorCh)
+		}(clientConn)
+	}
+	go func() {
+		workerGroup.Wait()
+		close(workersDone)
+	}()
+
+	if err := n.localStorage.Checkpointer().StartRestore(n.ctx, check); err != nil {
+		return checkpointStatusNext, fmt.Errorf("can't start checkpoint restore: %w", err)
+	}
+
+	// Prepare the heap of chunks.
+	chunks := &chunkHeap{
+		array: make([]*checkpoint.ChunkMetadata, len(check.Chunks)),
+		length: 0,
+	}
+	heap.Init(chunks)
+
+	for i, c := range check.Chunks {
+		heap.Push(chunks, &checkpoint.ChunkMetadata{
+			Version: 1,
+			Index: uint64(i),
+			Digest: c,
+			Root: check.Root,
+		})
+	}
+
+	// Feed the workers with chunks.
+	var next *checkpoint.ChunkMetadata
+	var outChan chan *checkpoint.ChunkMetadata
+
+	for {
+		if chunks.length == 0 {
+			next = nil
+			outChan = nil
+		} else {
+			next = heap.Pop(chunks).(*checkpoint.ChunkMetadata)
+			outChan = chunkDispatchCh
+		}
+
+		select {
+		case <-n.ctx.Done():
+			return checkpointStatusBail, n.ctx.Err()
+
+		case returned := <-chunkReturnCh:
+			if returned == nil {
+				// Restoration completed, no more chunks.
+				return checkpointStatusDone, nil
+			}
+			heap.Push(chunks, returned)
+
+		case status := <-errorCh:
+			return status, nil
+
+		case outChan <- next:
+			next = nil
+
+		case <-workersDone:
+			// No usable committee connections left, bail.
+			return checkpointStatusBail, storageClient.ErrStorageNotAvailable
+		}
+
+		if next != nil {
+			heap.Push(chunks, next)
+		}
+	}
+}
+
+func (n *Node) syncCheckpoint() error {
+	// Fetch metadata first from the current committee.
+	req := &checkpoint.GetCheckpointsRequest{
+		Namespace: n.commonNode.Runtime.ID(),
+	}
+	metadata, err := n.storageClient.GetCheckpoints(n.ctx, req)
```

Comment:
We should fetch checkpoints from multiple nodes and combine all of them. Otherwise this could end up with a bad storage node and you wouldn't be able to sync.